### ISSUE TITLE
systemd-networkd: configuring tunnels cause instability

### ISSUE THREAD

**Issue by lspgn** [systemd/systemd#26743]
### systemd version the issue has been seen with

251

### Used distribution

Ubuntu 22.10

### Linux kernel version used

5.19.0-35-generic

### CPU architectures issue was seen on

x86_64

### Component

systemd-networkd

### Expected behaviour you didn't see

After a restart of the `systemd-networkd` I was expecting to see the interfaces reconfigured working properly.

### Unexpected behaviour you saw

After a restart of the `systemd-networkd` service, if tunnels are configured (gre + sit), interfaces like Ethernet lost its IPv4 and IPv6 routes.

### Steps to reproduce the problem

Using Ubuntu 22.10 with systemd 251 (251.4-1ubuntu7.1).
I am originally using `netplan` and was able to reproduce the issue as long as there were tunnels in the config. `netplan generate` created the files listed below.

1. Create the following files

`/etc/systemd/network/15-sit.netdev`
```
[NetDev]
Name=foo
Kind=sit

[Tunnel]
Independent=true
Local=a.b.c.d
Remote=w.x.y.z
```

`/etc/systemd/network/15-sit.network`
```[Match]
Name=foo

[Network]
LinkLocalAddressing=ipv6
Address=a:b:c:d::1/64
ConfigureWithoutCarrier=yes
```

`/etc/systemd/network/15-bar.netdev`
```[NetDev]
Name=bar
Kind=gre

[Tunnel]
Independent=true
Local=a.b.c.d
Remote=w.x.y.z
```

`/etc/systemd/network/15-bar.network`
```[Match]
Name=bar

[Network]
LinkLocalAddressing=ipv6
Address=a:b:c:d::1/64
ConfigureWithoutCarrier=yes
```

2. Configure an Ethernet interface with IPv4, IPv6 (no DHCP)
3. Configure additional routes into `lo` interface
4. Restart the system
5. Everything should be configured and running normally
6. Run `sudo netplan --debug apply` or `systemctl restart systemd-networkd` at least once, wait for command to complete.

The `netplan` command will display something like:
```** (process:2271): DEBUG: 06:06:00.608: lo: adding new route
** (process:2271): DEBUG: 06:06:00.608: We have some netdefs, pass them through a final round of validation
** (process:2271): DEBUG: 06:06:00.608: eno1: setting default backend to 1
** (process:2271): DEBUG: 06:06:00.608: Configuration is valid
** (process:2271): DEBUG: 06:06:00.608: lo: setting default backend to 1
** (process:2271): DEBUG: 06:06:00.608: Configuration is valid
DEBUG:Merged config:
b''
Failed to reconfigure network interface eno1: Numerical result out of range
WARNING:Falling back to a hard restart of systemd-networkd.service
```
and sometimes will hang until cancelled to
```
subprocess.CalledProcessError: Command '['networkctl', 'reconfigure', '3', '1', '2', '5', '10', '9', '6', '8', '4', '7']' returned non-zero exit status 1.
```

And the logs of `systemd-networkd` will fill up with
```
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
Mar 10 05:54:57 sd-75303 systemd-networkd[2015]: Could not send netlink message: Numerical result out of range
```

There is a high chance static addresses and routes disappeared on the Ethernet interface. The service will keep failing and the presence or not of configuration files do not affect it.
Quickest remediation is to restart the system.

When trying without the 4 files and after a restart running
```bash
sudo ip tunnel add foo mode sit remote w.x.y.z local a.b.c.d
sudo ip -6 addr add a:b:c:d::1/64 dev foo
sudo ip link set foo up
sudo ip tunnel add bar mode gre remote w.x.y.z local a.b.c.d
sudo ip -6 addr add a:b:c:d::1/64 dev bar
sudo ip link set bar up
```

The service restarts and apply do not affect the interfaces.

Thank you for your help!

**Comment by yuwata** [comment]
Please create the following drop-in by `systemctl edit systemd-networkd.service`,
```
[Service]
Environment=SYSTEMD_LOG_LEVEL=debug
```
then restart systemd-networkd, and provide debugging logs obtained by something like the following:
```
journalctl -b --no-hostname -u systemd-networkd.service --no-pager > /tmp/log.txt
```

**Comment by yuwata** [comment]
BTW, some config files you provided does not have `[Match]` or `[NetDev]` section specifier...

**Comment by mickelmayers** [comment]
Hi,
It looks like we have similar issue. networkd only message in the logs is:
`Nov 29 14:57:24  systemd-networkd[10227]: ens2f0: Could not send rtnetlink message: Numerical result out of range`
Even in debug mode. 
Netplan when trying to try/apply:
`** (process:25537): DEBUG: 18:28:33.065: Configuration is valid
Failed to reload network settings: No such file or directory
WARNING:root:Falling back to a hard restart of systemd-networkd.service
Job for systemd-networkd.service failed because the control process exited with error code.`
The difference is we have ~921044 routes and networkd is throwing same message for each route.
The main problem is that random address form random interface is disappearing.
Any idea what could be wrong?

**Comment by yuwata** [comment]
Please provide debugging logs. Otherwise there is almost nothing we can do.

**Comment by yuwata** [comment]
> The difference is we have ~921044 routes and networkd is throwing same message for each route.

Do you manage the routes by networkd ? Or by another daemon? If these are not managed by networkd, then you should set ManageForeignRoutes=no in networkd.conf.

**Comment by mickelmayers** [comment]
We are using frr to managed routes. Which log you would like to see?

**Comment by mickelmayers** [comment]
In logs i see:
...
Nov 30 09:19:42 systemd-networkd[54866]: ens2f0: Remembering foreign route: dst: x.x.x.0/z, src: n/a, gw: y.y.y.y, prefsrc: n/a, scope: global, table: main(254), proto: bgp, type: unicast, nexthop: 41, priority: 20
Nov 30 09:19:42 systemd-networkd[54866]: ens2f0: Remembering foreign route: dst: c.c.c.c/23, src: n/a, gw: y.y.y.y, prefsrc: n/a, scope: global, table: main(254), proto: bgp, type: unicast, nexthop: 41, priority: 20
Nov 30 09:19:54 systemd-networkd[54866]: Enumeration completed
Nov 30 09:19:54 systemd[1]: Started Network Configuration.
Nov 30 09:19:55 systemd-networkd[54866]: ens2f0: Could not send rtnetlink message: Numerical result out of range
Nov 30 09:19:55 systemd-networkd[54866]: ens2f0: Could not send rtnetlink message: Numerical result out of range
Nov 30 09:19:55 systemd-networkd[54866]: ens2f0: Could not send rtnetlink message: Numerical result out of range
Nov 30 09:19:55 systemd-networkd[54866]: ens2f0: Could not send rtnetlink message: Numerical result out of range
Nov 30 09:19:55 systemd-networkd[54866]: ens2f0: Could not send rtnetlink message: Numerical result out of range
...
Nov 30 09:20:08 systemd-journald[1088]: Suppressed 1445694 messages from systemd-networkd.service
Nov 30 09:20:08 systemd-networkd[54866]: ens2f0: Removing route: dst: b.b.b.b/24, src: n/a, gw: y.y.y.y, prefsrc: n/a, scope: global, table: main(254), proto: bgp, type: unicast, nexthop: 41, priority: 20
Nov 30 09:20:08 systemd-networkd[54866]: ens2f0: Removing route: dst: d.d.d.d, src: n/a, gw: y.y.y.y, prefsrc: n/a, scope: global, table: main(254), proto: bgp, type: unicast, nexthop: 41, priority: 20

**Comment by mickelmayers** [comment]
ManageForeignRoutes=no has fixed the issue! Thank you!

**Comment by yuwata** [comment]
Hopefully fixed by #30268.
